core/server/models/user.js

```diff
@@ -355,16 +365,24 @@ User = ghostBookshelf.Model.extend({
      * Naive user add
      * Hashes the password provided before saving to the database.
      *
-     * @param {object} data
+     * We have to clone the data, because we remove values from this object.
+     * This is not expected from outside!
+     *
+     * @param {object} dataToClone
      * @param {object} options
      * @extends ghostBookshelf.Model.add to manage all aspects of user signup
      * **See:** [ghostBookshelf.Model.add](base.js.html#Add)
      */
-    add: function add(data, options) {
+    add: function add(dataToClone, options) {
         var self = this,
+            data = _.cloneDeep(dataToClone),
             userData = this.filterData(data),
             roles;
 
+        if (!userData.password) {
+            userData.password = utils.uid(50);
```

Comment:
hmm this does not work, defaults bookshelf call happens in a later order.
so our user add function does the following:

1. validate password length
2. hash password
3. call ghostBookshelf.Model.add and then defaults gets triggered

So that won't work.

I think the problem is that we are working not with event hooks.

It should work like it (didn't test)

1. ghostBookshelf.Model.add
2. defaults hook is triggered
3. validation is triggered (pwd length will be part of it)
4. creating event is triggered (hash pwd)